CityOfZion/neo-js
# Eliminate Circular references in neo-js

## Description

A major structural refactor orients around the goal of removing its existing circular reference architecture.

Changes:

* Upgrade `neon-js` from `2.1.1` to `3.2.1`.
* A new `Neo` class is introduced to act as 'Chain of Command'. `Node` is now a class aggregated by `Mesh` class.
* Rename `Neo` optional parameter `storage` to `storageOptions` and changed its structure:


```js

// Old:
const neo = new Neo({
  network: 'testnet',
  storage: {
    model: 'mongoDB',
    connectOnInit: true,
    connectionString: 'mongodb://localhost/sync_demo',
    collectionNames: {
      blocks: 'b_neo_t_blocks',
      transactions: 'b_neo_t_transactions',
      addresses: 'b_neo_t_addresses'
    }
  }
})

// New:
const neo = new Neo({
  network: 'testnet',
  storageOptions: {
    model: 'mongoDB',
    dataAccessOptions: {
      connectOnInit: true,
      connectionString: 'mongodb://localhost/sync_demo',
      collectionNames: {
        blocks: 'b_neo_t_blocks',
        transactions: 'b_neo_t_transactions',
        addresses: 'b_neo_t_addresses'
      }
    }
  }
})
```
* `Rpc` class is now using ES6 class syntax.
* Classes' default options are now in its own object instance `this.defaultOptions`.
* Move `profiles.js` to `/dist/common` directory.
* Remove all underscore prefix usages for method names.
* Rename classes to 'Pascal Case' and global variables to 'Camel Case'.
* Review all JSDocs. More work is still required tho.
* Add a regression test example: `example/nodes/regression-test.js` for semi-auto integrity check again local node.

Notes:
* JSDoc updates required, which will be as separate PR.

## Motivation and Context


To improve code readability and maintainability for the long run. This opens pave the way for Event Emitter (#95) and Deep-assign optional parameters.

## How Has This Been Tested?



Unit test and regression tests

## Types of changes:



## Checklist:
